src/devices/cpu/m6502/gew7.h

```diff
@@ -0,0 +1,93 @@
+// license:BSD-3-Clause
+// copyright-holders:Devin Acker
+/***************************************************************************
+
+    gew7.h
+
+    Yamaha GEW7, GEW7I, GEW7S (65c02-based)
+
+***************************************************************************/
+#ifndef MAME_CPU_M6502_GEW7_H
+#define MAME_CPU_M6502_GEW7_H
+
+#include "m6502mcu.h"
+#include "m65c02.h"
+#include "sound/gew7.h"
+
+class gew7_device : public m6502_mcu_device_base<m65c02_device>, public device_mixer_interface
+{
+public:
+
+	template<offs_t Num> auto port_in_cb() { return m_in_cb[Num].bind(); }
+	template<offs_t Num> auto port_out_cb() { return m_out_cb[Num].bind(); }
+
+protected:
+	gew7_device(const machine_config &mconfig, device_type type, bool ignore_in_ddr, const char *tag, device_t *owner, uint32_t clock);
+
+	virtual void device_add_mconfig(machine_config &config) override;
+
+	virtual void device_start() override;
+	virtual void device_reset() override;
+
+	virtual void internal_update(u64 current_time) override;
+	using m6502_mcu_device_base<m65c02_device>::internal_update;
+
+	u8 timer_stat_r();
+	void timer_stat_w(u8 data);
+	void timer_en_w(u8 data);
+
+	u8 timer_count_r(offs_t offset);
+	void timer_count_w(offs_t offset, u8 data);
+	u64 timer_update(int num, u64 current_time);
+
+	void bank_w(offs_t offset, u8 data);
+
+	u8 port_r(offs_t offset);
+	void port_w(offs_t offset, u8 data);
+	u8 port_ddr_r(offs_t offset);
+	void port_ddr_w(offs_t offset, u8 data);
+
+	void internal_map(address_map &map);
+
+	u8 m_timer_stat, m_timer_en;
+	u16 m_timer_count[2];
+	u64 m_timer_base[2];
+
+	devcb_read8::array<6> m_in_cb;
+	devcb_write8::array<6> m_out_cb;
+	u8 m_port_data[6];
+	u8 m_port_ddr[6];
+	// GEW7 and GEW7S seem to have different behavior when reading input bits
+	// (see gew7_device::port_r)
+	const bool m_ignore_in_ddr;
+
+	required_memory_region        m_rom;
+	required_memory_bank_array<2> m_bank;
+	u8 m_bank_mask;
+
+	required_device<gew7_pcm_device> m_pcm;
+};
+
+class ymw270f_device : public gew7_device
+{
+public:
+	ymw270f_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+};
+
+class ymw276f_device : public gew7_device
+{
+public:
+	ymw276f_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+};
+
+class ymw282f_device : public gew7_device
+{
+public:
+	ymw282f_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+};
+
+DECLARE_DEVICE_TYPE(YMW270F, ymw270f_device)
+DECLARE_DEVICE_TYPE(YMW276F, ymw276f_device)
+DECLARE_DEVICE_TYPE(YMW282F, ymw282f_device)
```

Comment:
What’s the difference between these three?  Right now they seem to just be aliases with different device types?  Different sound hardware?

Reply:
Right now the only functional difference is that YMW282F uses `m_ignore_in_ddr` for the alternate input port behavior. I had posted a comment earlier about `m_ignore_in_ddr` not being needed anymore, but that turned out to not be the case after more testing, so it's still there for now.

YMW276F can probably go away for now, it most likely works the same way as YMW270F (the PSR-110 service manual describes two different hardware revisions using both versions of the chip but the same program ROM for both).

Reply:
If it’s got open drain outputs and the DDR doesn’t cause it to read inputs from the output latch (i.e. the behaviour that Intel MCS-48 family MCUs typically use), you can use input code like:
```c++
u8 gew7_device::port_r(offs_t offset)
{
	const u8 out_data = m_port_data[offset] | ~m_port_ddr[offset];
	return out_data & m_in_cb[offset]();
}
```

Reply:
OK, that’s really nasty if you can overcome output drive in either direction.  Can you flip it around so rather than a read callback, the client calls a function to push the line state to the CPU?  That would allow you to have two bytes – one for the level, and one for whether it’s actively driven.